### ISSUE TITLE
[Safari] Tampermonkey cannot show installed userscripts

### ISSUE THREAD

**Issue by oscar-c** [Tampermonkey/tampermonkey#778]
## Expected Behavior
When switch to "Installed usercripts" tab, TM shows installed userscripts.

## Actual Behavior
The "Please wait..." prompt is there spinning forever, rendering the entire UI inoperable. 

## Specifications

- Safari: 13.0.2
- TM: 4.9.6037 (MAS Version)
- OS: macOS 10.15 Beta (19A582a) (the Golden Master)

AND

- Safari: 12.1.2
- TM: 4.9.6037 (MAS Version)
- OS: macOS 10.14.6

## Note

I get this behavor after I imported some backup from TM 4.7. The backup includes some customized script. 

And I want to know if there is a way to "factory reset" TM so I can investigate which specific userscript(s) caused the problem. I tried to uninstall/reinstall but the data is still there.

**Comment by derjanb** [comment]
> And I want to know if there is a way to "factory reset" TM so I can investigate which specific userscript(s) caused the problem.

Good point. I'll add a factory reset button to the host app's window. :)

You can reset the extension settings by executing this at a terminal:

```bash
rm -r Library/Containers/net.tampermonkey.*
```

**Comment by epsimatic** [comment]
Not opening another issue, but I have the problem triaged:

tl;dr: `{ "scripts": [ {"file_url": ...}]}` breaks import and script list

Steps to reproduce:

1. Reset Tampermonkey by deleting `~/Library/Containers/net.tampermonkey.SafariApp.Extension`
2. Start Safari
3. Open Tampermonkey preferences
4. Import file:
    a). [tampermonkey-backup-bad.txt](https://github.com/Tampermonkey/tampermonkey/files/3730545/tampermonkey-backup-bad.txt)
    b). [tampermonkey-backup-good.txt](https://github.com/Tampermonkey/tampermonkey/files/3730546/tampermonkey-backup-good.txt)
5. Confirm adding script
   - Expected behaviour: script imports, `-bad` and `good` files with minimal differences
   - Actual behaviour: `-good` backup loads, `-bad` backup hangs
6. Open Tampermonkey options, try open tabs.
   - Expected behaviour: able to open any tab, see the imported script in `scripts`
   - Actual behaviour: `-good` is good, `-bad` makes you unable to open `scripts` tab

Logs attached:

[log-good.txt](https://github.com/Tampermonkey/tampermonkey/files/3730557/log-good.txt)
[log-bad.txt](https://github.com/Tampermonkey/tampermonkey/files/3730559/log-bad.txt)

**Comment by epsimatic** [comment]
Temporary workaround: 

- reformat (lint) JSON file to a readable state
- remove `"file_url": ...` from every script 
- save and import

**Comment by oscar-c** [comment]
I've found the culprite, in my case at lease: https://greasyfork.org/en/scripts/34153-pixiv-plus

Regarless of being imported or installed freshly from greasyfork, it always causes infinite loading.

**Comment by paulzhang1992** [comment]
I tried to reset the extension settings. Installing any script will result in the loading situatio on installed script

**Comment by YuntianCheng** [comment]
> Temporary workaround:
> 
> * reformat (lint) JSON file to a readable state
> * remove `"file_url": ...` from every script
> * save and import

I have this issue but none of scripts I have installed has the "file_url" line

**Comment by oscar-c** [comment]
> > And I want to know if there is a way to "factory reset" TM so I can investigate which specific userscript(s) caused the problem.
> 
> Good point. I'll add a factory reset button to the host app's window. :)
> 
> You can reset the extension settings by executing this at a terminal:
> 
> ```shell
> rm -r Library/Containers/net.tampermonkey.*
> ```

Should be

 ```shell
 rm -r ~/Library/Containers/net.tampermonkey.*
 ```

**Comment by bengarland** [comment]
> I tried to reset the extension settings. Installing any script will result in the loading situatio on installed script

Same here. I tried installing a single script (Wanikani Override), which does not have file_url in it, and it will not load the Dashboard.

**Comment by cylott** [comment]
Same here. Happy to see the App store version and grateful it exists, but it doesn't work, at all. I tried to import my scripts and now it just spins and spins. I've tried the "factory reset" and deleting every folder and file it made in my ~/Library with no luck. Advice appreciated.

**Comment by epsimatic** [comment]
> Same here. Happy to see the App store version and grateful it exists, but it doesn't work, at all. I tried to import my scripts and now it just spins and spins. I've tried the "factory reset" and deleting every folder and file it made in my ~/Library with no luck. Advice appreciated.

Have you tried to remove `file_url` from backup file before import? https://github.com/Tampermonkey/tampermonkey/issues/778#issuecomment-542313292

**Comment by robbielj** [comment]
Removing `file_url` solves the issue for now, while the script cannot update via the file url.  Configuring update url of script after importing will cause the same "please wait" scenario.
hope there will be a fix soon.

**Comment by liyilin88923** [comment]
> I tried to reset the extension settings. Installing any script will result in the loading situatio on installed script

Same to me. I didn't import any previous userscripts but still encounter this problem once I install a new script. Since I didn't install the Pixiv Plus, so I guess other scripts also create this problem.

**Comment by robbielj** [comment]
> > I tried to reset the extension settings. Installing any script will result in the loading situatio on installed script
> 
> Same to me. I didn't import any previous userscripts but still encounter this problem once I install a new script. Since I didn't install the Pixiv Plus, so I guess other scripts also create this problem.

this is because the newly installed script has `file_url` field.
you may try copying a whole script into the tampermonkey editor and save, to see whether that will work.

**Comment by Chandler-Lu** [comment]
Use regex `("file_url".*?(user.js",))` to remove all `file_url` field.
And hope it will fix soon.

**Comment by liyilin88923** [comment]
> > > I tried to reset the extension settings. Installing any script will result in the loading situatio on installed script
> > 
> > 
> > Same to me. I didn't import any previous userscripts but still encounter this problem once I install a new script. Since I didn't install the Pixiv Plus, so I guess other scripts also create this problem.
> 
> this is because the newly installed script has `file_url` field. TM uses that for auto-updating scripts.
> you may try copying the whole script into the tampermonkey editor and save, to see whether that will work.

Thanks, I'll try that and see if it works.

**Comment by bengarland** [comment]
Dumb question, but where does Tampermonkey actually install the scripts on my Mac? They're not in ~/Library/Application Scripts/net.tampermonkey.SafariApp (or the related *.Extension folder). I've looked everywhere and can't find them. The Tampermonkey Options page is also no help.

**Comment by bengarland** [comment]
Here's what I did to fix the Dashboard loading error. I don't know if every step is necessary, but it worked:

1) Do NOT install scripts automatically by clicking an Install button on Greasyfork or wherever. Do NOT import scripts. Copy the raw js code for the script to your clipboard, then go to the Dashboard (it will load if you have zero scripts installed yet). Hit the + sign to add a new script. Paste the code into the editor. Go to File - Save To Disk in the editor window.

2) Then go to Actions column on the Dashboard, and select the Edit icon (paper with pencil) next to the script. This will bring you back to the Editor. Select the Settings tab. UN-check "Check for Updates" and click the Save button directly underneath it (this may not be necessary, but since someone else has mentioned issues with the file_url that points to where Tampermonkey checks for script updates, I thought it may be the issue).

3) Optional: If you have a script that needs to load first before all other scripts, make sure to change it to #1 in the Position drop-down menu at the very top of the script Settings page.

4) You can now close the Dashboard tab. Go to the Tampermonkey icon in your Safari toolbar and click it and select Dashboard. The Dashboard should now load properly.

**Comment by liyilin88923** [comment]
I'm glad to report that after I upgraded the tampermonkey to 4.9.6041 today from the Mac Appstore, the "installed userscripts" tab started to show normally. Problem solved! 

Thanks to the developer and all the others who helped!

**Comment by YuntianCheng** [comment]
> I'm glad to report that after I upgraded the tampermonkey to 4.9.6041 today from the Mac Appstore, the "installed userscripts" tab started to show normally. Problem solved!
> 
> Thanks to the developer and all the others who helped!

I have not received the update yet. And the App Store has a bug that some applications which has been installed on my Mac show a download logo in the Mac Store

**Comment by derjanb** [comment]
Yes, 6041 was reviewed and published this night. It should definitely fix

> I've found the culprit, in my case at least: https://greasyfork.org/en/scripts/34153-pixiv-plus

but I'm not sure about 

> tl;dr: { "scripts": [ {"file_url": ...}]} breaks import and script list

because this worked here fine all the time.

Can someone with the file_url issue and 6041 confirm whether it's working or not please? Thanks.

**Comment by liyilin88923** [comment]
> Yes, 6041 was reviewed and published this night. It should definitely fix
> 
> > I've found the culprit, in my case at least: https://greasyfork.org/en/scripts/34153-pixiv-plus
> 
> but I'm not sure about
> 
> > tl;dr: { "scripts": [ {"file_url": ...}]} breaks import and script list
> 
> because this worked here fine all the time.
> 
> Can someone with the file_url issue and 6041 confirm whether it's working or not please? Thanks.

I already got the 6041 and the installed scripts tab functioned normally after I install some scripts directly from the _greasyfork.org_

**Comment by liyilin88923** [comment]
> > I'm glad to report that after I upgraded the tampermonkey to 4.9.6041 today from the Mac Appstore, the "installed userscripts" tab started to show normally. Problem solved!
> > Thanks to the developer and all the others who helped!
> 
> I have not received the update yet. And the App Store has a bug that some applications which has been installed on my Mac show a download logo in the Mac Store

Maybe you could search the tampermonkey by name in the store window? Perhaps the update button will show there.

Plus, the Mac Appstore does have some weird problems sometimes, like the one you described. I guess the MAS just doesn't as good as the one on iOS.

**Comment by YuntianCheng** [comment]
> > > I'm glad to report that after I upgraded the tampermonkey to 4.9.6041 today from the Mac Appstore, the "installed userscripts" tab started to show normally. Problem solved!
> > > Thanks to the developer and all the others who helped!
> > 
> > 
> > I have not received the update yet. And the App Store has a bug that some applications which has been installed on my Mac show a download logo in the Mac Store
> 
> Maybe you could search the tampermonkey by name in the store window? Perhaps the update button will show there.
> 
> Plus, the Mac Appstore does have some weird problems sometimes, like the one you described. I guess the MAS just doesn't as good as the one on iOS.

Thanks, I find the update button by searching it, but tampermonkey did not upgrade at all after I clicked the update button. I think  I have to reboot my Mac now.

**Comment by epsimatic** [comment]
> Thanks, I find the update button by searching it, but tampermonkey did not upgrade at all after I clicked the update button. I think I have to reboot my Mac now.

Same issue here. It only updates successfully if Safari is closed at the moment of update. So, close Safari, update, launch Safari.

> Can someone with the file_url issue and 6041 confirm whether it's working or not please? Thanks.

I've updated, tried importing `tampermonkey-backup-bad.txt` from my report above — **Successful**

Imported my big backup — **Successful**

Issue is solved for me, Thank you Jan!

**Comment by derjanb** [comment]
😅

**Comment by YuntianCheng** [comment]
> > Thanks, I find the update button by searching it, but tampermonkey did not upgrade at all after I clicked the update button. I think I have to reboot my Mac now.
> 
> Same issue here. It only updates successfully if Safari is closed at the moment of update. So, close Safari, update, launch Safari.
> 
> > Can someone with the file_url issue and 6041 confirm whether it's working or not please? Thanks.
> 
> I've updated, tried importing `tampermonkey-backup-bad.txt` from my report above — **Successful**
> 
> Imported my big backup — **Successful**
> 
> Issue is solved for me, Thank you Jan!

Yes, I have updated successfully. I can finally enter the installed page.

**Comment by gabrielferrari714** [comment]
Issue is solved for me.
I have updated successfully. I can finally enter the installed page.
Thanks Jan. 🙏🏼

**Comment by gabrielferrari714** [comment]
> > > Thanks, I find the update button by searching it, but tampermonkey did not upgrade at all after I clicked the update button. I think I have to reboot my Mac now.
> > 
> > 
> > Same issue here. It only updates successfully if Safari is closed at the moment of update. So, close Safari, update, launch Safari.
> > > Can someone with the file_url issue and 6041 confirm whether it's working or not please? Thanks.
> > 
> > 
> > I've updated, tried importing `tampermonkey-backup-bad.txt` from my report above — **Successful**
> > Imported my big backup — **Successful**
> > Issue is solved for me, Thank you Jan!
> 
> Yes, I have updated successfully. I can finally enter the installed page.

Issue is solved here.
I have updated successfully (with Safari closed). Now I can enter the installed page.
Thank you so much, Jan. 🙏🏼
